javacPlugin/src/main/java/org/wpilib/javacplugin/CoroutineYieldInLoopDetector.java

```diff
@@ -0,0 +1,317 @@
+// Copyright (c) FIRST and other WPILib contributors.
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
+
+package org.wpilib.javacplugin;
+
+import com.sun.source.tree.CompilationUnitTree;
+import com.sun.source.tree.IdentifierTree;
+import com.sun.source.tree.LambdaExpressionTree;
+import com.sun.source.tree.MemberSelectTree;
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.source.tree.MethodTree;
+import com.sun.source.tree.WhileLoopTree;
+import com.sun.source.util.JavacTask;
+import com.sun.source.util.TaskEvent;
+import com.sun.source.util.TaskListener;
+import com.sun.source.util.TreeScanner;
+import com.sun.source.util.Trees;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import javax.lang.model.element.Element;
+import javax.lang.model.element.ExecutableElement;
+import javax.lang.model.element.VariableElement;
+import javax.lang.model.type.TypeMirror;
+import javax.tools.Diagnostic;
+
+/**
+ * Checks for {@code while} loops inside methods or lambda functions that accept coroutine
+ * arguments. If a loop does not call {@code yield()} on one of the most local coroutine objects, a
+ * compiler error will be emitted for that loop element. This check cannot be silenced.
+ */
+// Note: cannot be silenced because annotations cannot be placed on loops.
+// This is not legal Java:
+//   @SuppressWarnings("UnsafeCoroutineUsage")
+//   while (true) { ... }
+// Placing it at a higher level (lambda or method declaration) would silence ALL unsafe usage in
+// that expression; it's impossible to do on a case-by-case basis.
+public class CoroutineYieldInLoopDetector implements TaskListener {
+  private final JavacTask m_task;
+  private final Set<CompilationUnitTree> m_visitedCUs = new HashSet<>();
+
+  public CoroutineYieldInLoopDetector(JavacTask task) {
+    m_task = task;
+  }
+
+  @Override
+  public void finished(TaskEvent e) {
+    if (e.getKind() == TaskEvent.Kind.ANALYZE && m_visitedCUs.add(e.getCompilationUnit())) {
+      if (getCoroutineType() == null) {
```

Comment:
Lot of repetition here, esp around the command stuff

Perhaps extract some stuff to a base/utils class?